command/server/tls_util.go

```diff
@@ -0,0 +1,178 @@
+package server
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	"os"
+	"strings"
+	"time"
+)
+
+type CaCert struct {
+	PEM      string
+	Template *x509.Certificate
+	Signer   crypto.Signer
+}
+
+// GenerateCert creates a new leaf cert from provided CA template and signer
+func GenerateCert(caCertTemplate *x509.Certificate, caSigner crypto.Signer) (string, string, error) {
+	// Create the private key
+	signer, keyPEM, err := privateKey()
+	if err != nil {
+		return "", "", err
+	}
+
+	// The serial number for the cert
+	sn, err := serialNumber()
+	if err != nil {
+		return "", "", err
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return "", "", err
+	}
+
+	if hostname == "" {
+		hostname = "localhost"
+	}
+
+	// Create the leaf cert
+	template := x509.Certificate{
+		SerialNumber:          sn,
+		Subject:               pkix.Name{CommonName: hostname},
+		BasicConstraintsValid: true,
```

Comment:
As a complete nit, this could be removed (defaults to false) as a basic constraints aren't necessary on non-CA certs. This leaf isn't a CA (and `isCA` defaults to false in Go/x509), so it can be safely elided. The check in verification systems is always `hasBasicConstraint && basicConstraint.isCA == true` for CA-ness.

Reply:
Updated.